pubsys: replace rusoto with aws-sdk-rust

jpculp opened the pull request:
This is a work in progress to replace rusoto with aws-sdk-rust in pubsys.



**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

Commit:
infrasys: update to AWS SDK Rust

jpculp commented:
Added @rpkelly's infrasys commit with minimal adjustments for clippy and aws-sdk-for-rust versions.

jpculp commented:
Added logic to use specified profiles and endpoints in the AwsConfig.

zmrow commented on tools/infrasys/src/keys.rs:
```diff
@@ -89,9 +87,13 @@ impl KMSKeyConfigExt for KMSKeyConfig {
-                Region::from_str(region).context(error::ParseRegionSnafu { what: region })?,
-            );
+
+            let config = aws_config::from_env()
```
This is where I'd be the most worried and concerned.  In what order and precedence does the new SDK gather config/credentials compared to the old one?

jpculp replied:
My understanding is the default values for things that did not have an override are inserted at the `.load()` stage. If it's something that was overridden, but invalid (for example setting the region to a non-existent one) then the region will be left blank rather than receiving a default value.

zmrow commented on tools/infrasys/src/root.rs:
```diff
@@ -48,22 +60,26 @@ pub fn create_root(root_role_path: &Path) -> Result<()> {
     })?;
     fs::create_dir_all(role_dir).context(error::MkdirSnafu { path: role_dir })?;
+
+    let default_region = thread::spawn(move || {
```
Is there any reason to do this async?  Can we do it synchronously?  Spinning up runtimes in the middle of a program can get messy if other places are doing the same.

zmrow commented on tools/infrasys/src/root.rs:
```diff
@@ -40,6 +41,17 @@ pub fn check_root(root_role_path: &Path) -> Result<()> {
 }
 
+async fn async_get_region() -> Result<String> {
+    let default_region_fallback = "us-east-1";
```
I think I'd prefer to codify this as a constant somewhere so everything uses the same default region.  Nothing in the SDK provides anything like this?

jpculp replied:
The SDK has something called `RegionProviderChain` which we use in **testsys** to set the default to `us-west-2`. We could use that same approach for **infrasys**.

zmrow commented on tools/infrasys/src/s3.rs:
```diff
@@ -176,8 +179,11 @@ pub async fn upload_file(
 ) -> Result<()> {
-    let s3_client =
-        S3Client::new(Region::from_str(region).context(error::ParseRegionSnafu { what: region })?);
+    let config = aws_config::from_env()
```
Not sure if this throws everything for a loop, but is there a way to generate the `aws_config` in a single place and pass a reference around rather than continuously getting it?

jpculp replied:
That makes sense to me, but I'll have to look at it a bit more closely since this is @rpkelly's commit.

zmrow commented on tools/pubsys/src/aws/ami/register.rs:
```diff
@@ -74,17 +72,17 @@ async fn _register_image(
+                .set_volume_size(ami_args.root_volume_size)
+                .build(),
+        ))
+        .build();
```
I might look into what defaults the `build()` method adds to the struct compared to what Rusoto does in `Default::default()`... just to be sure nothing odd happens here.

zmrow commented on tools/pubsys/src/aws/ami/wait.rs:
```diff
@@ -11,8 +11,8 @@ use std::time::Duration;
     id: &str,
-    region: &Region,
-    sts_region: &Region,
+    region: &str,
```
Is there no stronger type we can use than `&str`?... I'd be nervous about passing invalid regions around.

jpculp replied:
Right now the region gets validated when the client configs are actually being generated (just before each client is created) so there shouldn't be a risk of a client getting an invalid region. I suspect the &str would be more efficient to pass around and then we don't have to convert it to a string at every error or debug message, but we could bring back and refactor the parse_region function if we want to make it a region object earlier.

The reason you're seeing these Region objects being replaced is they used to contain name and endpoint in a single object (because that's how rusoto handled endpoints) but now the endpoint is set separately and can just be fetched directly from pubsys-config's AwsConfig.

jpculp committed:
pubsys: update to AWS SDK Rust

jpculp commented:
Moved `build_client_config()` into `client.rs` to make it easier to compare and fixed logic where applicable. Chaining of assumed roles is commented out because it's missing something.

jpculp commented:
Further revisions and discussions on upstream PR 2415.
